The `cfunction` 1.0 deserves

vtjnash opened the pull request:
This PR turns `cfunction` into a macro, and has it work more similarly to `ccall` (no runtime compiler required, defined by the containing method parameters, supports any type object including closures). In most cases, v0.6 code should just need to add an `@`. But it also has new advanced capabilities to efficiently create gc-managed runtime closures pointers! This will fix #12256, fix #26070 and close #1096 (I feel slightly cheated to not get to close a 3-digit issue on this one 😜).

Implements https://github.com/vtjnash/julep/wiki/0004:-Static-compilation-for-cfunction

The best place to see the net effect of this change is perhaps the stdlib, but also the docs (where I have an example of qsort), or the Julep (which has a more complete exploration of qsort usages), or lastly perhaps the tests (which try to have examples of weird uses of almost everything).

~~Syntax questions:~~

1. ~~Creating a closure over the value of a local variable requires annotating with `$`: i.e. `ptr = @cfunction $cl RT (A, B)`. However, I've also considered instead using `local` for this purpose instead: `ptr = @cfunction local cl RT (A, B)`. Thoughts on which one reads better? As can be seen from there being zero usages in this repo (outside of tests), this feature isn't a particularly common need / the other form is more preferred and easier to use (when possible), so I don't really care either way.~~

2. ~~Should I continue to use basically the same syntax as before (i.e. just add `@`), or use the following alternate syntax rearrangement, or just try to provide both: `ptr = @cfunction cl(A, B)::RT` (which, for the variant that creates a closure over the value of the local variable `cl`, would be `ptr = @cfunction $cl(A, B)::RT` or `ptr = @cfunction local cl(A, B)::RT`)~~

vtjnash committed:
cfunction macro: extend cfunction capabilities

Provide static support for handling dynamic calls and closures

maleadt commented on stdlib/LibGit2/src/tree.jl:
```diff
@@ -12,27 +12,33 @@ subtree, traversing upwards through it, then traversing the next right subtree
+function treewalk(f, tree::GitTree, post::Bool = false)
     # NOTE: don't use @check/GitError directly, because the code can be arbitrary
+    payload = Any[ tree, f ]
+    cbf = @cfunction(function treewalk_callback(root_cstr::Cstring, entry_ptr::Ptr{Cvoid}, payload::Vector{Any})::Cint
```
Isn't it possible to capture `tree` and `f` instead of using the `payload` mechanism?

vtjnash replied:
Yes, it's just slower that way than to just use the C API correctly

StefanKarpinski replied:
I presume it's slower because closure support requires a stub that sets the locals and jumps to the "real" function? Whereas without any closed-over variables, it can just call directly.

vtjnash replied:
Yes, and in addition to the indirect function call, it also requires allocation to be able to track the lifetime, and runtime work to manage and handle that, etc.

vtjnash committed:
eval: add fast-path for `.` expr

macro head uses this to resolve symbols,
which we want to allow even in pure context (generated functions)

JeffBezanson commented on base/c.jl:
```diff
@@ -16,24 +16,56 @@ respectively.
+    @cfunction(callable, ReturnType, (ArgumentTypes...,)) -> Ptr{Cvoid}
+    @cfunction(\$callable, ReturnType, (ArgumentTypes...,)) -> CFunction
+
+Generate a C-callable function pointer from the Julia function `f`
```
There is no `f` :)

JeffBezanson commented on base/exports.jl:
```diff
@@ -925,6 +925,7 @@ export
 
 # C interface
     cfunction,
```
Remove this?

JeffBezanson commented on doc/src/manual/calling-c-and-fortran-code.md:
```diff
@@ -911,6 +917,34 @@ mylibvar = Libdl.dlopen("mylib")
+## Closure cfunctions
+
+The first argument to [`@cfunction`](@ref) can be marked with a `$`, in which case
+the return value will instead be a `struct CFunction` which closes over the argument.
```
Out of curiosity: what happens if I don't use a `$`, but I give a global name that's bound to a non-singleton function object?

vtjnash replied:
Like `ccall`, it's resolved by the macro to close over the value (not the binding), since that's almost always what you want anyways, and makes it easier to avoid accidentally shooting your self in the (gc) foot.

JeffBezanson commented:
I think the `$` syntax is good. It looks like the first argument can also be a function expression, which also gets `$` treatment?

vtjnash commented:
> which also gets $ treatment?

Not really. Because I felt that it read nicely to keep the function definition and use together this way, I'm (ab)using the knowledge that it got evaluated in global scope early, and then treated as a constant. In either case, it can be an arbitrary expression, the `$` keyword just indicates when to evaluate it.

stevengj commented:
+1 for `ptr = @cfunction $cl(A, B)::RT`

nalimilan commented on doc/src/manual/calling-c-and-fortran-code.md:
```diff
@@ -182,26 +185,26 @@ an integer less/greater than zero if `a` should appear before/after `b` (or zero
 using the `qsort` function (rather than Julia's built-in `sort` function). Before we worry about
 calling `qsort` and passing arguments, we need to write a comparison function that works for some
-arbitrary type T:
+arbitrary type T (which defines `<`):
```
`T` isn't used in the example anymore.

nalimilan commented on doc/src/manual/calling-c-and-fortran-code.md:
```diff
@@ -182,26 +185,26 @@ an integer less/greater than zero if `a` should appear before/after `b` (or zero
 
 Notice that we have to be careful about the return type: `qsort` expects a function returning
-a C `int`, so we must be sure to return `Cint` via a call to `convert` and a `typeassert`.
+a C `int`, so we annotate the return type of the function be sure it returns a `Cint`.
```
Missing "to"?

tknopp commented:
@vtjnash: Any chance that this will fix https://discourse.julialang.org/t/performance-issue-with-gtk-jl/6171. That inference on callbacks requires up to minutes is a major issue when using Gtk currently.

giordano commented:
Did I miss it or this change isn't mentioned in `NEWS.md`?

vtjnash commented on base/c.jl:
```diff
@@ -16,24 +16,56 @@ respectively.
+    f::Any
+    _1::Ptr{Cvoid}
+    _2::Ptr{Cvoid}
+    let construtor = false end
```
It disables creation of the default constructors (ensuring that this type is not accidentally constructible from user code).
